academic_observatory_workflows/workflows/crossref_events_telescope.py

```diff
@@ -12,383 +12,859 @@
+    if cursor is not None:
+        query_params["cursor"] = cursor
+
+    req = requests.models.PreparedRequest()
+    req.prepare_url(url, query_params)
+    return req.url
+
+
+class EventRequest:
+    def __init__(self, action: str, day: pendulum.DateTime, mailto: str):
+        self.action = action
+        self.day = day
+        self.mailto = mailto
+
+    @property
+    def cursor_file_name(self):
+        return f"{self.action}-{self.day.format(DATE_FORMAT)}-cursor.txt"
+
+    @property
+    def data_file_name(self):
+        return f"{self.action}-{self.day.format(DATE_FORMAT)}.jsonl"
+
+    def make_url(self, rows: int = 1000, cursor: str = None) -> str:
+        """Make a URL for this EventsRequest. See documentation for more details:
+
+        Example queries: https://www.eventdata.crossref.org/guide/service/query-api/#example-queries
+        Edited events: See documentation for more details: https://www.eventdata.crossref.org/guide/service/query-api/#edited-events
+        Deleted events: https://www.eventdata.crossref.org/guide/service/query-api/#deleted-events
+
+        :param rows: the number of rows.
+        :param cursor: the cursor.
+        :return: the URL.
         """
-        release.transform()
+
+        return make_crossref_events_url(
+            action=self.action, start_date=self.day, end_date=self.day, mailto=self.mailto, rows=rows, cursor=cursor
+        )
+
+    def __str__(self):
+        return f"EventsRequest('{self.action}, {self.day.format(DATE_FORMAT)}, {self.mailto}')"
+
+    def __repr__(self):
+        return f"EventsRequest('{self.action}, {self.day.format(DATE_FORMAT), {self.mailto}}')"
 
 
-def parse_event_url(url: str) -> Tuple(str, str):
-    """Parse the URL to get the event type and date
+def make_day_requests(start_date: pendulum.DateTime, end_date: pendulum.DateTime, mailto: str) -> List[DayRequest]:
+    """Create the upsert and delete URLs. The interval for dates is [start_date, end_date). On the first run only
+    create events are returned in the upsert_events list, on subsequent runs, create and edit event URLs are populated
+    into upsert_events and delete events in delete_events.
 
-    :param url: The url
-    :return: The event type and date
+    :param start_date: the start date. Inclusive.
+    :param end_date: the end date. Exclusive.
+    :param mailto: the mailto email address.
+    :return: the URLs.
     """
-    event_type = url.split("?mailto")[0].split("/")[-1]
-    if event_type == "events":
-        date = url.split("from-collected-date=")[1].split("&")[0]
-    else:
-        date = url.split("from-updated-date=")[1].split("&")[0]
 
-    return event_type, date
+    events = []
+    period = pendulum.period(start_date, end_date)
+    for day in period.range("days"):
+        if day != end_date:
+            events.append(DayRequest(day, mailto))
+    return events
+
 
+def download_events(request: EventRequest, download_folder: str, n_rows: int):
```

Comment:
Codecov says that the first two 'if' statements aren't covered in tests. It might be worth making sure they're included as the logic flow here is pretty important.

Reply:
Yeah sure

Reply:
I've added in another test as well to check `fetch_events` as well, with the real endpoints.